watertap/unit_models/ion_exchange_0D.py

```diff
@@ -0,0 +1,1346 @@
+        self.t_rinse = Var(
+            initialize=360, units=pyunits.s, bounds=(120, 1200), doc="Rinse time"
+        )
+
+        self.main_pump_power = Var(
+            initialize=4e-6, units=pyunits.kW, doc="Main pump power"
+        )
+
+        self.regen_pump_power = Var(
+            initialize=4e-6, units=pyunits.kW, doc="Regen pump power"
+        )
+
+        self.bw_pump_power = Var(
+            initialize=4e-6, units=pyunits.kW, doc="Backwash pump power"
+        )
+
+        self.rinse_pump_power = Var(
+            initialize=4e-6, units=pyunits.kW, doc="Rinse pump power"
+        )
+
+        self.pump_efficiency = Var(
+            initialize=0.8,
+            units=pyunits.dimensionless,
+            bounds=(0, 1),
+            doc="Pump efficiency",
+        )
+
+        tmp_dict = dict(**self.config.property_package_args)
+        tmp_dict["has_phase_equilibrium"] = False
+        tmp_dict["parameters"] = self.config.property_package
+        tmp_dict["defined_state"] = True  # inlet block is an inlet
+
+        self.properties_in = self.config.property_package.state_block_class(
+            self.flowsheet().config.time,
+            doc="Material properties of inlet",
+            default=tmp_dict,
+        )
+
+        # Add outlet and waste block
+        tmp_dict["defined_state"] = False  # outlet and waste block is not an inlet
+        self.properties_out = self.config.property_package.state_block_class(
+            self.flowsheet().config.time,
+            doc="Material properties of outlet",
+            default=tmp_dict,
+        )
+
+        self.properties_regen = self.config.property_package.state_block_class(
+            self.flowsheet().config.time,
+            doc="Material properties of waste",
+            default=tmp_dict,
+        )
+
+        # Add ports - oftentimes users interact with these rather than the state blocks
+        self.add_port(name="inlet", block=self.properties_in)
+        self.add_port(name="outlet", block=self.properties_out)
+        self.add_port(name="waste", block=self.properties_regen)
+
+        # Add constraints
+        # =========== EQUILIBRIUM ===========
+
+        @self.Constraint(
+            ion_set,
+            doc="Separation factor calc",
+        )
+        def eq_sep_factor(b, j):
+            return b.separation_factor[j] == 1 / b.selectivity[j]
+
+        @self.Constraint(
+            ion_set,
+            doc="Langmuir isotherm",
+        )
+        def eq_langmuir(b, j):
+            if j == target_ion:
+                return b.separation_factor[j] == (
+                    b.c_norm[j] * (1 - b.resin_eq_capacity / b.resin_max_capacity)
+                ) / (b.resin_eq_capacity / b.resin_max_capacity * (1 - b.c_norm[j]))
+            else:
+                return Constraint.Skip
+
+        @self.Constraint(doc="Resin equilibrium concentration bounds")
+        def eq_resin_eq_capacity(b):
+            return b.resin_eq_capacity <= b.resin_max_capacity
```

Comment:
I am willing to wager that you needed this inequality to prevent poor convergence that would have come from the manner in which the above 2 constraints are formulated. 

Theoretically, your `eq_capacity` should be able to be equal to your `max_capacity`, but if you let that happen then `separation_factor` in the above constraint goes to zero, which then causes your `selectivity` variable in the constraint above that to need to go towards infinity. 

I think a lot of this could be avoided if you reformulate the first 2 constraints.

Reply:
I added this constraint because I was getting valid solves where `resin_eq_capacity` was greater than `resin_max_capacity`, which shouldn't happen.